sysbuild: Import image devicetrees

57300 opened the pull request:
Make the image devicetrees available to sysbuild for advanced use cases.

This is done by calling `zephyr_dt_import` on each image target. The DT
target properties are all expected to be prefixed with `DT_`, so there
should be no conflicts.

Thus, the DT contents of a given `<image>` can be retrieved using e.g.:

    dt_nodelabel(<var> NODELABEL <label> TARGET <image>)

as noted in `extensions.cmake`.

nordicjm reviewed (dismissed):
Very useful addition!

bjarki-andreasen commented:
Could we add these additions to the docs somewhere? Looking at the current docs, only reference to these macros seems to be in an old release note, it would be great to actually have these macros (including the existing dt_ ones) somewhere :) until now, I have to look to the scripts to figure out what we support

57300 commented:
> Could we add these additions to the docs somewhere? Looking at the current docs, only reference to these macros seems to be in an old release note, it would be great to actually have these macros (including the existing dt_ ones) somewhere :) until now, I have to look to the scripts to figure out what we support

I don't know how many CMake extensions are currently documented, but it seems to be that most aren't. I think it might be nice to first consider creating some API doc auto-generation for CMake (or finding an existing solution) to avoid maintaining multiple copies of the same text. Maybe @tejlmand knows more about this?

bjarki-andreasen commented:
> > Could we add these additions to the docs somewhere? Looking at the current docs, only reference to these macros seems to be in an old release note, it would be great to actually have these macros (including the existing dt_ ones) somewhere :) until now, I have to look to the scripts to figure out what we support
> 
> I don't know how many CMake extensions are currently documented, but it seems to be that most aren't. I think it might be nice to first consider creating some API doc auto-generation for CMake (or finding an existing solution) to avoid maintaining multiple copies of the same text. Maybe @tejlmand knows more about this?

Maybe we can add some sample which uses the new extensions? or at least test cases for it?

57300 commented:
> Maybe we can add some sample which uses the new extensions? or at least test cases for it?

I think adding unit tests would make sense, maybe similar to the ones I had written for `zephyr_get()`. Right now I don't have the bandwidth for that though.

bjarki-andreasen commented:
> > Maybe we can add some sample which uses the new extensions? or at least test cases for it?
> 
> I think adding unit tests would make sense, maybe similar to the ones I had written for `zephyr_get()`. Right now I don't have the bandwidth for that though.

ok, I would like to have some tests or similar before the PR is merged, I have added the DNM label until then :)

nordicjm commented:
@57300 why is this DNM?
@tejlmand please review

57300 commented:
> @57300 why is this DNM?

@bjarki-andreasen marked it because he came in and asked for tests: https://github.com/zephyrproject-rtos/zephyr/pull/73903#issuecomment-2358032482

I had agreed initially, but I would first need to add tests for the regular `dt_*` API, which deserves a separate PR.

pdgendt commented:
Needs a rebase

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -763,3 +765,86 @@ function(sysbuild_images_order variable dependency_type)
+      )
+    endif()
+
+    cmake_language(CALL ${name} ${var} ${SBDT_UNPARSED_ARGUMENTS})
```
this is bound to break at some point.

Forwarding arguments blindly will not work in all cases, so when using functions like `cmake_language()` and / or forwarding arguments then we really need to have full control / knowledge how this is being called.

This proposal can lead to unexpected surprises / hard to debug issues.

Please also take a look at: https://crascit.com/2019/01/29/forwarding-command-arguments-in-cmake/

57300 replied:
Removed.

tejlmand reviewed (changes requested):
Thanks for this proposal.

Having devicetrees from images available is really needed and the proposal appreciated.
There are some comments related to some of the suggested ideas.

57300 commented on cmake/modules/extensions.cmake:
```diff
@@ -3891,6 +3898,10 @@ endfunction()
 # PROPERTY <prop> : The alias to check
 # REQUIRED        : Generate a fatal error if the alias is not found
 function(dt_alias var)
+  if(NOT DEFINED DEVICETREE_TARGET OR NOT TARGET "${DEVICETREE_TARGET}")
+    message(FATAL_ERROR "dt_alias(${ARGV0} ...) devicetree is not available.")
+  endif()
```
Good point. I can't remember why I convinced myself that using a parent scope variable was okay.

If it's acceptable to add a `TARGET` argument to all `dt_*` functions, then that would practically remove the need for `sysbuild_dt_*` functions, since one would be free to use e.g.:
```cmake
# in sysbuild:
dt_nodelabel(var TARGET image NODELABEL label)

# instead of:
sysbuild_dt_nodelabel(var IMAGE image NODELABEL label)
```

If that feels right to you, then that would also remove the complexity involved in forwarding arguments from `sysbuild_dt_*` functions, which I didn't get right the first time: https://github.com/zephyrproject-rtos/zephyr/pull/73903#discussion_r2091405265

tejlmand replied:
> If it's acceptable to add a `TARGET` argument to all `dt_*` functions, then that would practically remove the need for `sysbuild_dt_*` functions, since one would be free to use e.g.:

I would say it is.
It's very similar to other functions where we allow flexibility by using custom targets for scoping, for example `import_kconfig`.
Or you can say `zephyr_get()` indirectly uses target, although less visible because the SYSBUILD flag.

> If that feels right to you, then that would also remove the complexity involved in forwarding arguments from sysbuild_dt_* functions, which I didn't get right the first time: https://github.com/zephyrproject-rtos/zephyr/pull/73903#discussion_r2091405265

Sounds like a win-win :+1:

57300 commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -512,6 +512,8 @@ function(ExternalZephyrProject_Cmake)
   endif()
   load_cache(IMAGE ${ZCMAKE_APPLICATION} BINARY_DIR ${BINARY_DIR})
   import_kconfig(CONFIG_ ${BINARY_DIR}/zephyr/.config TARGET ${ZCMAKE_APPLICATION})
+  set(DEVICETREE_TARGET ${ZCMAKE_APPLICATION}_devicetree_target)
+  include(${BINARY_DIR}/zephyr/dts.cmake)
```
I didn't think much of it, because our CMake modules also expect certain variables to be set beforehand, which are documented as "inputs", so it's not like this is a bad practice for CMake files in general.

It's not quite so blatant there as it is here, though, and maybe dealing with a generated file also feels different.

I'm skeptical of going the `import_kconfig()` route, because that function parses an actual `.config` file, whereas an `import_dt()` couldn't feasibly take an actual `.dts` file, much less an `edt.pickle`. We could at least replace the generated `dts.cmake` with a JSON file or something, and use it to populate target properties, as long as we don't have to rework the whole CMake DT solution from the ground up.

tejlmand replied:
> I'm skeptical of going the import_kconfig() route, because that function parses an actual .config file, whereas an import_dt() couldn't feasibly take an actual .dts file, much less an edt.pickle, so it would feel asymmetrical.

The intention was not to go exactly the Kconfig route but to have a function for loading devicetree.
A function which build system can call, and which would allow adjusting how content is loaded. Similar to how you can request `import_kconfig()` to use a different prefix or populate a specific target.

So minimum request is to have a function call where the caller decides how to load and adjust content.
Could be `import_dts()`.
Having such call will also allow to rework the solution any way we like without impacting callers of `import_dts()`

> We could at least replace the generated dts.cmake with a JSON file or something, and use it to populate target properties, as long as we don't have to rework the whole CMake DT solution from the ground up.

Now you're bring up the topic of changing the format of the python generated output file :wink: .

If we are to adjust, then json could be an option, or we could consider aligning with zephyr_modules.py which makes semicolon lists like this:
```
"DTS_ROOT":"/projects/github/ncs/modules/hal/adi"
```
or:
```
"nrf":"/projects/github/review/nrf":"/projects/github/review/nrf"                                   
```
but if content loading is handled through a function, then we can change any way we like in future.
But I do consider such re-write outside scope of this PR, unless you have time and believe it to be a good idea.

Note, the comment: `Perhaps generating a CMake dts file from python is not the best way if we need to do such hacks.` was more an observation and not a request to do such big changes in this PR.

> because our CMake modules also expect certain variables to be set beforehand

True, and it would be nice to find the time to do a followup to #41301 so that Zephyr's CMake modules became even cleaner, while keeping current re-usability.

57300 replied:
Thanks for clarifying.

> Now you're bring up the topic of changing the format of the python generated output file 😉 .

I'm just thinking out loud about how to apply your suggestion and make it architecturally sound :)

I guess there's no need to change the format in this PR. I only considered it because I imagined that `import_dts()` would only be able to take `dts.cmake` as input, which is very esoteric, so that didn't sit right with me.

Now I'm considering moving this Python invocation:
https://github.com/zephyrproject-rtos/zephyr/blob/6e34acc4e296b44b0b2142ba56480a6f2ee08e38/cmake/modules/dts.cmake#L368-L374

to encapsulate it in the new CMake function and take `edt.pickle` as input:

```cmake
function(import_edt file)
  if(${file} IS_NEWER_THAN ${file}.cmake)
    # run python script
  endif()
  set(DEVICETREE_TARGET devicetree_target)
  include(${file}.cmake)
endfunction()
```

There would be a byproduct `edt.pickle.cmake`, which could be loaded multiple times without rerunning the script, as long as the corresponding `edt.pickle` doesn't change.

> Having such call will also allow to rework the solution any way we like without impacting callers of `import_dts()`

I propose calling the function `import_edt()`, so that the API stability would largely be tied to edtlib.

tejlmand replied:
> I only considered it because I imagined that import_dts() would only be able to take dts.cmake as input, which is very esoteric

I agree with you there.
Was mostly trying to see how we could improve PR without requesting a major restructure / update.

> Now I'm considering moving this Python invocation:
> ...
> to encapsulate it in the new CMake function and take edt.pickle as input:

Having the function being able to take `edt.pickle` as input makes the function much cleaner seen from the caller's perspective.

Thanks for that proposal :+1: 

> I propose calling the function import_edt(), so that the API stability would largely be tied to edtlib.

:+1: for that.

57300 replied:
Added as `zephyr_dt_import(EDT_PICKLE_FILE ...)` to fit into an existing `extensions.cmake` section.

57300 committed:
sysbuild: Import image devicetrees

Make the image devicetrees available to sysbuild for advanced use cases.

This is done by calling `zephyr_dt_import` on each image target. The DT
target properties are all expected to be prefixed with `DT_`, so there
should be no conflicts.

Thus, the DT contents of a given `<image>` can be retrieved using e.g.:

    dt_nodelabel(<var> NODELABEL <label> TARGET <image>)

as noted in `extensions.cmake`.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
cmake: modules: Add zephyr_dt_import

Extract the part of `dts.cmake` that invokes `gen_dts_cmake.py`, then
generalize it into a CMake extension, which can be reused by sysbuild.

The Python script itself is also updated, so that the generated CMake
file can accept an input variable DEVICETREE_TARGET, which comes from
the `zephyr_dt_import(TARGET ...)` argument.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
cmake: extensions: Add TARGET arguments to the dt_* API

The CMake dt_* API is backed by target properties. By default, it uses
the target named `devicetree_target` that is initialized in `dts.cmake`.

We should be able to customize this, now that we have a function for
loading those properties into a different target. For example:

    zephyr_dt_import(EDT_PICKLE_FILE /path/to/edt.pickle TARGET target)

    dt_nodelabel(node NODELABEL label TARGET target)

This would make it possible to reference multiple devicetrees in the
same build system, particularly sysbuild.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
